transform-plugins/src/main/java/io/cdap/plugin/XMLParser.java

```diff
@@ -77,65 +80,79 @@ public XMLParser(Config config) {
+    // Get failure collector for updated validation API
+    FailureCollector collector = getContext().getFailureCollector();
+    outSchema = config.getOutputSchema(collector);
+    xPathMapping = getXPathMapping(collector);
+    collector.getOrThrowException();
   }
 
   /**
    * Valid if xpathMappings and schema contain the same field names.
    */
-  private void validateXpathAndSchema() throws UnsupportedEncodingException {
-    xPathMapping = getXPathMapping();
+  private void validateXpathAndSchema(FailureCollector collector) {
+    xPathMapping = getXPathMapping(collector);
     List<Schema.Field> outFields = outSchema.getFields();
     // Checks if all the fields in the XPath mapping are present in the output schema.
     // If they are not a list of fields that are not present is included in the error message.
-    StringBuilder notOutput = new StringBuilder();
     for (Schema.Field field : outFields) {
       String fieldName = field.getName();
       if (!xPathMapping.keySet().contains(field.getName())) {
-        notOutput.append(fieldName + ";");
+        ValidationFailure failure = collector.addFailure(
+          String.format("Field '%s' must be present in XPath Mapping.", fieldName), null);
+        if (field.getSchema().isNullableSimple()) {
+          String typeString = field.getSchema().getNonNullable().getType().toString().toLowerCase();
+          failure.withConfigElement(Config.FIELD_TYPE_MAPPING, String.format("%s:%s", fieldName, typeString));
+        } else {
+          // This should never happen assuming all mappable types are nullable simple but just in case
+          failure.withConfigProperty(Config.FIELD_TYPE_MAPPING);
+        }
       }
     }
-    if (notOutput.length() > 0) {
-      throw new IllegalArgumentException("Following fields are not present in output schema :" +
-                                           notOutput.toString());
-    }
   }
 
-  private Map<String, String> getXPathMapping() throws UnsupportedEncodingException, IllegalArgumentException {
+  private Map<String, String> getXPathMapping(FailureCollector collector) {
     Map<String, String> map = new HashMap<>();
     String[] xpaths = config.xPathFieldMapping.split(",");
     for (String xpath : xpaths) {
       String[] xpathmap = xpath.split(":"); //name:xpath[,name:xpath]*
 
       if (xpathmap.length != 2) {
-        throw new IllegalArgumentException("Invalid XPath mapping specified : '" + xpath + "'");
-      }
-
-      if (xpathmap[0] == null || xpathmap[0].trim().isEmpty()) {
-        throw new IllegalArgumentException("XPath mapping is missing a field name : '" + xpath + "'");
-      }
-
-      if (xpathmap[1] == null || xpathmap[1].trim().isEmpty()) {
-        throw new IllegalArgumentException("XPath mapping is missing xpath : '" + xpath + "'");
+        collector.addFailure(String.format("Invalid XPath mapping: '%s'.", xpath), null)
+            .withConfigElement(XPATH_MAPPINGS, xpath);
```

Comment:
nit: indentation

Reply:
Fixed, also I switched to the proper CDAP IntelliJ config so hopefully these shouldn't happen anymore